pkg/policy/mapstate.go

```diff
@@ -195,29 +255,46 @@ func (keys MapState) addKeyWithChanges(key Key, entry MapStateEntry, adds, delet
 
 // deleteKeyWithChanges deletes a 'key' from 'keys' keeping track of incremental changes in 'adds' and 'deletes'.
 // The key is unconditionally deleted if 'cs' is nil, otherwise only the contribution of this 'cs' is removed.
-func (keys MapState) deleteKeyWithChanges(key Key, cs CachedSelector, adds, deletes MapState) {
+func (keys MapState) deleteKeyWithChanges(key Key, owner MapStateOwner, adds, deletes MapState) {
 	if entry, exists := keys[key]; exists {
-		if cs != nil {
+		if owner != nil {
 			// remove the contribution of the given selector only
-			if _, exists = entry.selectors[cs]; exists {
+			if _, exists = entry.owners[owner]; exists {
 				// Remove the contribution of this selector from the entry
-				delete(entry.selectors, cs)
-				// key is not deleted if other selectors still need it
-				if len(entry.selectors) > 0 {
+				delete(entry.owners, owner)
+				owner.RemoveDependent(keys, key)
+				// key is not deleted if other owners still need it
+				if len(entry.owners) > 0 {
 					return
 				}
 			} else {
-				// 'cs' was not found, do not change anything
+				// 'owner' was not found, do not change anything
 				return
 			}
 		}
+
+		// Remove this key from all owners' dependents maps if no owner was given
```

Comment:
This is the case this is handling, right? Would it help provide better context to readers if we added a small extra comment here?

```suggestion
		// Remove this key from all owners' dependents maps if no owner was given
		// This is typically the case if the key is being removed due to a deny rule being added.
```

Reply:
Actually I'm not sure if dependent rules are ever deleted without passing the owner, added this for completeness so that if that ever happens there will be do dangling references left behind.

Reply:
When I searched for the caller of this function, I found some calls which explicitly passed `nil` as the `MapStateOwner` parameter so I think this is actually being used. That's where I came up with the comment suggestion.

Reply:
Added this to the comment here:
```
		// Owner is nil when deleting more specific entries (e.g., L3/L4) when
		// adding deny entries that cover them (e.g., L3-deny).
```